### PR TITLE
Update to React v0.14

### PR DESCRIPTION
**Pull request by radiodario**
woah, big one. Added React v0.14, updated react-router to 1.0.0-rc3, refactored to the newer react-router API, fixed tests, removed warnings (where possible). There was a couple of warnings that i cannot fix because they come from upstream (in alt) and they are a known issue that will be fixed in v0.17.5 (but will be a breaking change).

[read more about the alt issue here](https://github.com/goatslacker/alt/issues/492)
[read more about react v0.14 here](https://facebook.github.io/react/blog/2015/10/07/react-v0.14.html)

### CODE REVIEW

**Comment by peter-mouland** [comment]
@radiodario i guess this is still Work In Progress? 
..http://ft-editorial-lantern-pr-123.herokuapp.com/articles/02287f44-2a3d-11e5-8613-e7aedbb7bdb7 doesn't want to load

**Comment by radiodario** [comment]
I Swear It Worked FIne In My Local Copy lol

**Comment by radiodario** [comment]
seems to be some issue with the minified version - not sure what

**Comment by radiodario** [comment]
We can't upgrade until Alt fixes their breaking change, so we should probably freeze this for the time being.

File: src/server/server.js [hunk 1]
```diff
@@ -57,11 +58,11 @@ app.use('/bye', (req, res) =>{
 app.use('/', authRouter);
 app.use('/api/v0', ensureApiAuthenticated, apiRouter);
 app.use('/', ensureAuthenticated, dataPreloader);
-app.use('/', ensureAuthenticated, function appRouter(req, res) {
-  renderRoute(routes, req, res)
+app.use('/', ensureAuthenticated, function appRouter(req, res, next) {
+  renderRoute(routes, req, res, next);
```
Comment (i-like-robots) [inline src/server/server.js]: next?

File: src/server/server.js [hunk 2]
```diff
@@ -57,11 +58,11 @@ app.use('/bye', (req, res) =>{
...
 });
 app.use(function ErrorHandler(err, req, res, next) {
-  renderRoute(errorRoutes, req, res)
+  renderRoute(errorRoutes, req, res, next);
```
Comment (i-like-robots) [inline src/server/server.js]: next?

File: src/server/server.js [hunk 3]
```diff
@@ -74,6 +75,8 @@ var server = app.listen(config.port, function () {
 });
 
 function renderRoute(route, req, res) {
+  res.set('Link', prefetch.join(','));
+  res.locals.data = res.locals.data || {};
```
Comment (i-like-robots) [inline src/server/server.js]: not sure these should be here

File: src/shared/components/Search.js [hunk 4]
```diff
@@ -15,7 +13,6 @@ import ArticleQueryActions from '../actions/ArticleQueryActions';
 import ComparatorQueryActions from '../actions/ComparatorQueryActions';
 
 import _ from 'underscore';
-import moment from 'moment';
```
Comment (i-like-robots) [inline src/shared/components/Search.js]: These deps might need to stay

File: src/shared/components/SectionHeadlineStats.js [hunk 5]
```diff
@@ -10,22 +10,21 @@ function getMetrics (config, data, comparatorData) {
...
   let colWidth = 12 / keys.length;
 
-  keys.forEach(function (value) {
+  keys.forEach(function (value, i) {
     let componentConfig = config[value];
     componentConfig.metric = data[value];
     componentConfig.comparatorMetric = comparatorData[config[value].comparatorFormatName] || undefined;
-    componentConfig.comparatorName = data[value] || '';
-
+    componentConfig.comparatorName = value || '';
```
Comment (i-like-robots) [inline src/shared/components/SectionHeadlineStats.js]: probably should stay the same here

File: src/shared/controllers/AppController.js [hunk 6]
```diff
@@ -1,23 +1,33 @@
...
     return (<DocumentTitle title='Lantern'>
-      <AltContainer store={UserStore} >
-        <User />
-        <NavBar />
+      <div>
+        <User {...this.props} />
+        <NavBar {...this.props} />
         {this.props.children}
-      </AltContainer>
+      </div>
```
Comment (i-like-robots) [inline src/shared/controllers/AppController.js]: Stay as `AltContainer`?

File: src/shared/components/SectionHeadlineStats.js (hunk 5 above)
Comment (radiodario) [inline src/shared/components/SectionHeadlineStats.js]: this was causing a warning in react... i'll make a note of this and change it back if we see a bug

File: src/shared/controllers/AppController.js (hunk 6 above)
Comment (radiodario) [inline src/shared/controllers/AppController.js]: that's how it is on my local copy, how weird

File: webpack.config.prod.js [hunk 7]
```diff
@@ -18,7 +18,7 @@ module.exports = {
   },
   plugins: [
     new webpack.optimize.DedupePlugin(),
-    new webpack.optimize.UglifyJsPlugin({output: {comments: false}, minimize: true, compress: { warnings: false }}),
+    //    new webpack.optimize.UglifyJsPlugin({output: {comments: false}, minimize: true, compress: { warnings: false }}),
```
Comment (radiodario) [inline webpack.config.prod.js]: this should not be commented out - had done it to try and debug unminified code on prod

**Pull request closed without merging**